web-api/src/middleware/call_context.rs

```diff
@@ -0,0 +1,220 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::{future::Future, sync::Arc};
+
+use actix_web::{
+    body::BoxBody,
+    dev::{Service, ServiceRequest, ServiceResponse},
+    web::Data,
+    HttpMessage,
+    HttpRequest,
+};
+use anyhow::anyhow;
+use futures_util::{
+    future::{self, Either},
+    FutureExt,
+};
+use reqwest::header::HeaderValue;
+use serde::{Deserialize, Serialize};
+use tokio::task::futures::TaskLocalFuture;
+use tracing::{info_span, trace, Instrument};
+use uuid::Uuid;
+
+use crate::error::early_failure::middleware_failure;
+
+pub(crate) struct ApiCallContext {
+    #[allow(unused)]
+    pub(crate) tenant_id: TenantId,
+    pub(crate) request_id: RequestId,
+}
+
+tokio::task_local! {
+    static API_CALL_CONTEXT: Arc<ApiCallContext>;
+}
```

Comment:
I originally thought we need this for the db handling but it turns out we do not.

Still it does:

- remove all the hard to understand "request id" injection type magic
- mages the (follow up) db code much simpler (saving 500+ line changes or so)

There is no drawback (with how we do things) for the error handling usage, but there is
one drawback for the Database usage, `tokio::spawn` doesn't carry over the context so
if we want to clone storage and pass it to a spawned task and use it there we need to take
additional care to also carry over the context, this is currently not the case anywhere in our 
code and writing a `spawn_with_context` function isn't too hard.

Given how much task locals makes things simpler I decided to go with it for now and we might
later one change the DB handling to work more seamless with spawn if we  need it.

Reply:
I just realized I have to implement the more complex DB version earlier (due to the details of the connection pools `before_acquire` error handling). I still would keep it in task local storage anyway.